### ISSUE TITLE
ERC: Contract Ownership Governance Standard

### ISSUE THREAD

**Issue by zemse** [ethereum/EIPs#2766]
Opening this thread to discuss a construction that can be standardized.

Traditionally, many contracts that require that they be owned or controlled in some way use `EIP-173` which standardizes the use of ownership in the smart contracts. For example to withdraw funds or perform administrative actions.

```solidity
contract dApp {
  function doSomethingAdministrative() external onlyOwner {
    // admin logic that can be performed by a single wallet
  }
}
```

Often, such administrative rights for a single wallet are written for maintainance purpose, but it overpowers the owner wallet and users need to trust the owner. Rescue operations by owner wallet have raised questions on decentralised nature of the dApp projects. Also, there is a possibility of compromise of owner's private key. All it needs is a single wallet to do a single transaction in a dApp to do misuse of the powers.

This can be solved by having the `dApp` smart contract to consider address of another smart contract `Governance` as it's owner (instead of a wallet address).

```solidity
contract Governance {
  function makeGovernedCall(
    uint256 _nonce,
    address _to,
    bytes memory _data,
    bytes[] memory _signatures
  ) external payable {
    // performs initial checks like enough signatures

    // once all checks are done, the governance makes
    //  an internal call to the application contract
    (bool _success, ) = _to.call{value: msg.value}(_data);
  }
}
```

Calls can be made by this contract to `EIP-173` compatible comtracts. This means existing contracts with a centralized wallet controlling the administrative ownership can be decentralized to multiple wallets with minimum complicated setup.

The reference implementation is available [here](https://github.com/zemse/smart-contract-governance/blob/master/contracts/SimpleGovernance/Governance.sol) in the [repository](https://github.com/zemse/contract-ownership-governance#contract-ownership-governance). I have also written some test cases in the same repo.

This is still very nascent. I believe improvising this would benefit users of existing and future dApps on Ethereum by not requiring to trust owners of dApps. Also do any similar constructions exists?

**Comment by zemse** [comment]
Multi sig wallets do have the same functionality. My bad, closing it.

**Comment by mudgen** [comment]
Hey, this is not just useful for wallets.  I think this would be very useful for governance of diamonds.  It would decentralized the ability to add new functions to diamonds.   See the Diamond Standard here: https://github.com/ethereum/EIPs/issues/2535

**Comment by zemse** [comment]
Thanks for your response. This draft was focused on implementation and while I was going through 0xProject contracts, I found that they already implemented the idea [here](https://github.com/0xProject/0x-monorepo/blob/development/contracts/multisig/contracts/src/ZeroExGovernor.sol) kind of 2-3 years ago. Also, looks like they have on-chain governance, however the reference example I've included was based on off-chain governance.

>  I think having a standard specifically for decentralizing the ownership of contracts is a good idea.

Now I realise that this EIP should be more focused on a consistent interface. This would enable tools like EtherScan can display governance information about a dApp smart contract (like number of owners and required consensus). This can definitely update facets of a diamond proxy implementation with decentralized control. Including some community tested codes (with minimum to no change required) with the EIP will help more developers to easily implement this in their dApp projects requiring ownership functionality.

Reopening to do a rewrite of the current EIP draft.

**Comment by mudgen** [comment]
Hi @zemse,

Yes, this is great!

**Comment by mudgen** [comment]
I like the latest version.  I like that there is an interface for on-chain voting and off-chain voting.

**Comment by zemse** [comment]
@mudgen Thanks for the fixing some stuff! The updated the description above. Would love any suggestions for making this better.

Tomorrow I'll try to update the reference implementation with the latest version of the interfaces.

**Comment by mudgen** [comment]
@zemse I think it makes sense to start contacting the developers of existing DAOs and people who may be interested in this standard and getting feedback from them. I can help with this and I'll start doing this.

**Comment by fubuloubu** [comment]
Took a quick look, I think `required() -> (uint256, uint256)` could be simplified a bit logically if it where split into two functions, something like `threshold() -> uint256` (number of positive votes needed) and  `quorum() -> uint256` (number of total votes needed). You could further specify that `threshold <= quorum` in well-formed implementations.

An implementation should specify that a vote "passes" if `totalVotes >= quorum && positiveVotes >= threshold`. The implementation might also say a vote may be denied if `totalVotes >= quorum && negativeVotes >= threshold`. There might be further extensions for time periods, lack of quorum, etc. There are tons of ways you can implement alterations, but I think having `threshold` and `quorum` make the intent behind `required` much clearer than contextually parsing the denominator.

**Comment by zemse** [comment]
Hi @fubuloubu, I think for the check `totalVotes >= quorum && positiveVotes >= threshold`, the contract should have functionality for governor to able to vote negative (please correct me if I'm wrong). But most of the implementations that I have seen have their governors two options: either case a vote (positive only) or ignore by doing nothing (considered negative by default). What do you think, governor ignoring works fine or does enabling negative votes significantly enhances a governance contract?

Also most implementations had a fixed `required` value (some of them: [0xproject](https://github.com/0xProject/0x-monorepo/blob/development/contracts/multisig/contracts/src/MultiSigWallet.sol#L34), [starkex](https://github.com/starkware-libs/starkex-contracts/blob/d6bde00ba6f5fa38c05e2e2222ddc6a2eaf2c562/scalable-dex/contracts/src/committee/Committee.sol#L11), [jointerinc](https://github.com/jointerinc/jointer-token/blob/90256da08d0e2248c4367df09eeb1b97e015da10/contracts/common/MultiSigGovernance.sol#L48), [qdao](https://github.com/platinum-engineering/qdao-smartcontract-qdao/blob/2b52adce88e54b0d009555baf042a0d7f50e4a03/contracts/Access/Multiownable.sol#L7), [melonproject](https://github.com/melonproject/melon/blob/master/contracts-old/MultiSigWallet.sol#L27)). But if more governors are added, the required number of votes still remains the same. There is no way in such implementations to instead have `50%+` or `66%+` consensus like ([posdao](https://github.com/poanetwork/posdao-contracts/blob/master/contracts/ValidatorSetAuRa.sol#L360)). I think it could be helpful if it's possible to have both, i.e. fixed votes as well as % votes and let governors to switch to % or back to fixed as desired during the course. Does this make sense?

I agree that `required() -> (uint256, uint256)` can be simplified and a split into two methods could do that. For e.g. `required() => 2` and `base() => 3` means `66%+` consensus and if governors set `base` to `0` it means `required` is fixed. Any name suggestions or alternative ways of doing this?

**Comment by fubuloubu** [comment]
Yes, depending on the implementation, you don't have to have negative voting. I was just using `postiveVotes` to make my point, it could just be `votes` (and it doesn't matter because it's an internal variable anyways).

The point of an ERC should be allow a standardized way of interacting with what could be multiple iterations of the same idea: simplified voting, timebound rejection, negative votes, etc. A successful ERC doesn't need to dictate behavior of what the implementation chooses to do with the result, only allows for a range of possibilities prescribed by the intent of the standard. To say it simply, the standard should describe what things mean and how to work with them, and leave some grey area for people to experiment with. Too loose, and the standard becomes brittle and hard to use. Too tight, and no one wants to use it because it's too specific.

We're mostly saying the same thing, I'm just hoping my reasoning can help you in thinking about designing your standard.

---

I don't think `base` is a very descriptive name, perhaps something like `minimum` would be more descriptive as to what you mean by returning a parameter that dictates what the minimum quorum is required for the vote to pass. Also, having `base` shift to 0 _depending_ on the use case I think will lead this standard to being poorly implemented. That's why I was trying to describe a simple invariant that is valid in all cases (`votesCast >= quorum && votesFor >= threshold`), which is something that is easier to ensure people implement correctly when they implement this standard. I mean, people still forget to implement the ERC20 return boolean sometimes, so being explicit and straightforward helps prevent this kind of misuse.

**Comment by mudgen** [comment]
@fubuloubu @zemse I agree it is good to make things very clear and explicit in the standard.

Another important point concerning designing the standard is tooling. I agree that it is good for a standard to provide flexibility in its implementation, but it shouldn't be so flexible that it becomes less useful for tools.  Tools that can automate or fill in the boilerplate of a standard can be very useful, but the standard has to standardize enough specifics.  Specifically a standard should standardize the interfaces, which is what this standard is doing which is good.

**Comment by fulldecent** [comment]
Please add full text to issue and make a PR to put draft in the repo. This is the correct processes for EIPs.

**Comment by zemse** [comment]
I've updated the full text to this issue and opened the PR #2767. We've also made some changes to make methods explicit based on the feedback received.

**Comment by mudgen** [comment]
There has been some feedback in the pull request which was merged: https://github.com/ethereum/EIPs/pull/2767

**Comment by fulldecent** [comment]
Good luck with this, I have made a review

# Review of current version

## Style

- Reference and use RFC 2119 at the top of the specification
- Add Solidity versions inline to specification a la https://eips.ethereum.org/EIPS/eip-721#specification
- Fix "off chain" -> "off-chain"
- Refer to "ERC-173" as "ERC-173 (DRAFT)", it is not yet finalized

## BC

- Upgrade ERC-165 compliance from "should" to "MUST" in all cases

- The features `GovernorPowerUpdated`, `powerOf`, `totalPower`, `setGovernor`, `GovernorPowerUpdated` are all entirely duplicative of ERC-20 and other subsequent tokens. Either replace these and simply make ERC-20 (or other) a dependency of this ERC or explain that this design consideration was considered and provide a strong reason why this backwards compatibility is abandoned.

## Design

- Reorder to save a storage slot: 

    ```solidity
     struct Transaction {
        uint256 votes;
        bool executed; // This gets combined with address below
        address destination;
        uint256 value;
        bytes data;
    }
    ```

- `createTransaction` should not confirm. As is, utility of standard is limited to cases where governors will propose transactions. A more general, and better, approach is to allow implements to have anybody propose transactions.

- Off-chain `executeTransaction` is missing `value` argument

- NatSpec for off-chain `executeTransaction` signature format is underspecified

- `Only Governance can call` on `executeTransaction` is over specified 
## Naming

- Rename `TransactionCreated` -> `TransactionProposed`
- `event TransactionConfirmed(uint256 indexed transactionId);` should include which governor
- Rename `revokeConfirmation` -> `revokeTransaction`
- Alternatively, consider: `voteForTransaction(uint256 _transactionId)` and `voteAgainstTransaction(uint256 _transactionId)` because this can be more clear

## Security

- As-is, implementations will require for-loops to implement `setGovernor`, `getTransaction`, `executeTransaction`, `confirmTransaction` and/or `revokeConfirmation`. This creates possibility for a denial of service. This is not explained or mitigated in the paper.

## Related

- In case you are interested, 0xcert has implemented something similar, you can see it at https://github.com/0xcert/framework/blob/4cfca4559e58c8bef08b82501757e75a37dc7a91/packages/0xcert-ethereum-gateway-contracts/src/contracts/actions-gateway.sol

**Comment by zemse** [comment]
@fulldecent Thank you for highlighting important points. Some names were inspired from [MultiSigWallet](https://github.com/ConsenSys/MultiSigWallet/blob/master/MultiSigWalletWithDailyLimit.sol#L362-L364) too with an attempt to keep familiar naming. Missed mentioning them, will do that.

**Comment by mudgen** [comment]
@fulldecent Could a section be added to EIP-1 that contains the RFC 2119 mention and text.  And have it say that RFC 2119 applies to all EIPs?  That way every EIP doesn't have to include it.

**Comment by fulldecent** [comment]
No, we cannot have EIPs reference EIP-1 because that means that EIPs will no longer be free standing. Plus it is a good reminder to standards authors that they need to use precise language when they have to copy the RFC 2119 boiler plate directly into their text.

---

But separately, EIP-1 should be updated to recommend that all EIPs will reference on RFC 2119 (rather than me just bringing it up each time during review). This is just helpful because a lot of people here do not have experience with standards work and that is just a good idea.

I have given up on proposing "material" changes to EIP-1 because there is no project scope on this repository, and therefore no mechanism to accept changes. And the only "review" is a bunch of random people that do not have commit access. I don't have time to join committees that are ineffectual.

**Comment by MicahZoltu** [comment]
Someone could add the RFC 2119 to the EIP template.  That would probably go over smoother than trying to change EIP-1.

**Comment by zemse** [comment]
The draft has been updated recently. As a changelog along with their reasons:

- **Removed specification for Transaction execution**: This allows dev to follow any other onchain/offchain standard or custom implementation as per their needs. For e.g. negative voting https://github.com/ethereum/EIPs/issues/2766#issuecomment-657166432
- **Use ERC-20 as a dependency**: As highlighted in https://github.com/ethereum/EIPs/issues/2766#issuecomment-660714769, The features GovernorPowerUpdated, powerOf, totalPower, setGovernor, GovernorPowerUpdated are all entirely duplicative of ERC-20 and other subsequent tokens. This allows to the Governance contract to contain ERC-20 compatible code or seperate out the token functionality in a seperate contract like Compound's GovernerAlpha architecture.
- **Changed `required() -> (uint256,uint256)` to `quorumVotes() -> uint256`**: The required() returning two numbers was a bit confusing method as highlighted in https://github.com/ethereum/EIPs/issues/2766#issuecomment-657166432. Now this allows devs to keep it simple or do some custom logic under it.

Thanks to @mudgen, @fubuloubu, @fulldecent, @MicahZoltu for taking time to review and highlighting problems and giving feedback for the initial draft.

I've updated the new draft text in this issue's description. Requesting review for this. Also, the draft is much smaller than initial version. Looking to fix any problems in the specification, so that this standard becomes more clear and useful to developers for implementing.
